include/xtensor/xnpy.hpp

```diff
@@ -0,0 +1,666 @@
+
+            if ((magic_string_length + 2 + 4 + header_length) % 16 != 0)
+            {
+                // TODO: display warning
+            }
+
+            char* buf = new char[header_length];
+            istream.read(buf, header_length);
+            std::string header(buf, header_length);
+            delete[] buf;
+
+            return header;
+        }
+
+        struct npy_file
+        {
+            npy_file() = default;
+
+            npy_file(std::vector<std::size_t>& shape, bool fortran_order,
+                     std::string typestring)
+                : m_shape(shape), m_fortran_order(fortran_order), m_typestring(typestring)
+            {
+                // Allocate memory
+                m_word_size = (std::size_t) atoi(&typestring[2]);
+                m_n_bytes = compute_size(shape) * m_word_size;
+                m_buffer = new char[m_n_bytes];
+            }
+
+            ~npy_file()
+            {
+                if (m_buffer != nullptr)
+                {
+                    delete m_buffer;
+                }
+            }
+
+            template <class T, layout_type L>
+            auto cast_impl(bool check_type)
+            {
+                if (m_buffer == nullptr)
+                {
+                    throw std::runtime_error("This npy_file has already been cast.");
+                }
+                T* ptr = reinterpret_cast<T*>(&m_buffer[0]);
+                std::vector<std::size_t> strides(m_shape.size());
+                std::size_t sz = compute_size(m_shape);
+
+                // check if the typestring matches the given one
+                if (check_type && m_typestring != detail::build_typestring<T>())
+                {
+                    throw std::runtime_error("formatting error: template format not matching"s + m_typestring + 
+                                             " vs "s + detail::build_typestring<T>());
+                }
+
+                if ((L == layout_type::column_major && !m_fortran_order) ||
+                    (L == layout_type::row_major && m_fortran_order))
+                {
+                    throw std::runtime_error("fortran order and xarray layout type do not match up.");
+                }
+
+                compute_strides(m_shape,
+                                m_fortran_order ? layout_type::column_major : layout_type::row_major,
+                                strides);
+                std::vector<std::size_t> shape(m_shape);
+
+                return std::make_tuple(ptr, sz, std::move(shape), std::move(strides));
+            }
+
+            template <class T, layout_type L = layout_type::dynamic>
+            auto cast(bool check_type = true) &&
+            {
+                auto cast_elems = cast_impl<T, L>(check_type);
+                m_buffer = nullptr;
+                return std::move(xadapt(std::move(std::get<0>(cast_elems)), std::get<1>(cast_elems),
```

Comment:
I think that there is no need to call `std::move` on `xadapt` since xadapt should already return an rvalue.